src/other/tables.jl

```diff
@@ -0,0 +1,48 @@
+using Tables, IteratorInterfaceExtensions
+
+Tables.istable(::Type{<:AbstractDataFrame}) = true
+Tables.columnaccess(::Type{<:AbstractDataFrame}) = true
+Tables.columns(df::AbstractDataFrame) = df
+Tables.rowaccess(::Type{DataFrame}) = true
+Tables.rows(df::DataFrame) = eachrow(df)
+
+Tables.schema(df::AbstractDataFrame) = Tables.Schema(names(df), eltypes(df))
+
+function _isiterabletable(x::T) where {T}
+    Base.isiterable(T) || return false
+    return Base.IteratorEltype(x) === Base.HasEltype() ? eltype(x) <: NamedTuple : false
+end
+
+function DataFrame(x::T) where {T}
+    if Tables.istable(T)
+        columns = Tables.columns(x)
+        return DataFrame([collect(u) for u in Tables.eachcolumn(columns)],
+                          collect(Tables.schema(columns).names))
+    end
+    y = IteratorInterfaceExtensions.getiterator(x)
+    yT = typeof(y)
+    if Base.isiterable(yT)
+        if Base.IteratorEltype(yT) === Base.HasEltype() && eltype(y) <: NamedTuple
+            columns = Tables.buildcolumns(Tables.Schema(eltype(y)), y)
+            return DataFrame([collect(u) for u in columns],
+                        collect(Tables.schema(columns).names))
+        else
+            # non-NamedTuple or UnknownEltype
+            columns = Tables.buildcolumns(Tables.Schema(propertynames(first(x)), nothing), y)
```

Comment:
Hm, I think the call to ``first`` here is quite problematic. For some query operators, iterating the first element is costly (i.e. the whole query essentially is executed for that first iteration), so we really want to avoid calling ``iterate`` twice on a query. I think that would happen here: it would once get called for the ``first`` call, and then again later for the materialization, right? I think with this implementation, essentially for many queries we would essentially execute them twice to materialize them into a ``DataFrame``...

Reply:
I also noticed this at https://github.com/JuliaData/Tables.jl/pull/10#discussion_r214523261.